src/super_context.ml

```diff
@@ -95,52 +132,336 @@ let expand_ocaml_config t pform name =
+            | Some p ->
+              let x =
+                Pkg_version.read sctx p >>^ function
+                | None   -> [Value.String ""]
+                | Some s -> [String s]
+              in
+              Resolved_forms.add_ddep acc ~key x
+            | None ->
+              Resolved_forms.add_fail acc { fail = fun () ->
+                Loc.fail loc
+                  "Package %S doesn't exist in the current project." s
+              }
+          end
+        | Macro (Read, s) -> begin
+            let path = Path.relative dir s in
+            let data =
+              Build.contents path
+              >>^ fun s -> [Value.String s]
+            in
+            Resolved_forms.add_ddep acc ~key data
+          end
+        | Macro (Read_lines, s) -> begin
+            let path = Path.relative dir s in
+            let data =
+              Build.lines_of path
+              >>^ Value.L.strings
+            in
+            Resolved_forms.add_ddep acc ~key data
+          end
+        | Macro (Read_strings, s) -> begin
+            let path = Path.relative dir s in
+            let data =
+              Build.strings path
+              >>^ Value.L.strings
+            in
+            Resolved_forms.add_ddep acc ~key data
+          end
+        | Macro (Path_no_dep, s) -> Some [Value.Dir (Path.relative dir s)])
+    in
+    Option.iter res ~f:(fun v ->
+      acc.sdeps <- Path.Set.union
+                     (Path.Set.of_list (Value.L.deps_only v)) acc.sdeps
+    );
+    res
+
+  let with_expander sctx ~dir ~dep_kind ~scope ~targets_written_by_user
+        ~map_exe ~bindings ~f =
+    let acc = Resolved_forms.empty () in
+    ( f (expander ~acc sctx ~dir ~dep_kind ~scope ~targets_written_by_user ~map_exe ~bindings)
+    , acc
+    )
+end
+
+let map_exe sctx =
+  match sctx.host with
+  | None -> (fun exe -> exe)
+  | Some host ->
+    fun exe ->
+      match Path.extract_build_context_dir exe with
+      | Some (dir, exe) when dir = sctx.context.build_dir ->
+        Path.append host.context.build_dir exe
+      | _ -> exe
+
+let expand_and_eval_set t ~scope ~dir ?(bindings=Pform.Map.empty) set ~standard =
   let open Build.O in
-  let parse ~loc:_ s = s in
-  let (syntax, files) =
-    let f = expand_vars_path t ~scope ~dir ?bindings in
-    Ordered_set_lang.Unexpanded.files set ~f in
-  let f = expand_vars t ~mode:Many ~scope ~dir ?bindings in
-  match Path.Set.to_list files with
+  let ((partial, paths), resolved_forms) =
+    Expander.with_expander t ~dir ~dep_kind:Required
+      ~scope ~targets_written_by_user:(Static [])
+      ~map_exe:(map_exe t)
+      ~bindings
+      ~f:(fun f ->
+        let f : Ordered_set_lang.Unexpanded.expander =
+          { f = fun ~mode sw ->
+              String_with_vars.partial_expand ~mode ~dir ~f sw
+          } in
+        Ordered_set_lang.Unexpanded.expand set ~dir ~f)
+  in
+  let f =
+    Expander.Resolved_forms.build resolved_forms >>^ fun dynamic_expansions ->
```

Comment:
This is supposed to be called only once. Additionally, `resolved_forms` is mutable, so this should only be called once all the mutations have been done.

Reply:
It is called only once. Also, all the mutations occurred in the call to `with_expander` above it.